state-chain/pallets/cf-lending-pools/src/general_lending/general_lending_tests.rs

```diff
@@ -5261,3 +5283,147 @@ fn must_have_refund_address_for_loan_asset() {
+
+			LoanAccounts::<Test>::insert(BORROWER, loan_account.clone());
+		})
+		// Running the next block will run upkeep and trigger liquidation. We just want to make sure
+		// this does not panic.
+		.then_process_next_block()
+		.then_execute_with(|_| {
+			// We expect to find a the loan account in liquidation but with no swaps associated with
+			// it. This means it is stuck like this until we fix it. This is acceptable behavior
+			// because this state should be unreachable.
+			assert_eq!(
+				LoanAccounts::<Test>::get(BORROWER).unwrap().liquidation_status,
+				LiquidationStatus::Liquidating {
+					liquidation_type: LiquidationType::Hard,
+					liquidation_swaps: BTreeMap::new()
+				}
+			)
+		});
+}
+
+#[test]
+fn same_asset_loan() {
+	const EXPANDED_PRINCIPAL: AssetAmount = PRINCIPAL / 2;
+	const SWAP_DEFICIT: AssetAmount = 1_000;
+	const SWAP_OUTPUT_AMOUNT: AssetAmount = INIT_COLLATERAL + PRINCIPAL - SWAP_DEFICIT;
+
+	new_test_ext()
+		.with_funded_pool(INIT_POOL_AMOUNT)
+		.execute_with(|| {
+			set_asset_price_in_usd(LOAN_ASSET, 1);
+			MockBalance::credit_account(&BORROWER, LOAN_ASSET, INIT_COLLATERAL);
+			MockLpRegistration::register_refund_address(BORROWER, LOAN_CHAIN);
+
+			// Should be able to create a loan where the loan asset is the same as the collateral
+			// asset
+			assert_eq!(
+				LendingPools::new_loan(
+					BORROWER,
+					LOAN_ASSET,
+					PRINCIPAL,
+					Some(LOAN_ASSET),
+					BTreeMap::from([(LOAN_ASSET, INIT_COLLATERAL)])
+				),
+				Ok(LOAN_ID)
+			);
+
+			// And expand using the loan asset we just got to make sure it works.
+			assert_ok!(LendingPools::expand_loan(
+				RuntimeOrigin::signed(BORROWER),
+				LOAN_ID,
+				EXPANDED_PRINCIPAL,
+				BTreeMap::from([(LOAN_ASSET, PRINCIPAL)])
+			));
+
+			// Now start a liquidation
+			assert_ok!(LendingPools::initiate_voluntary_liquidation(RuntimeOrigin::signed(
+				BORROWER
+			)));
+		})
+		.then_process_next_block()
+		.then_execute_with(|_| {
+			// Check that the liquidation is happening
+			assert_eq!(
+				MockSwapRequestHandler::<Test>::get_swap_requests(),
+				BTreeMap::from([(
+					SwapRequestId(0),
+					MockSwapRequest {
+						input_asset: LOAN_ASSET,
+						output_asset: LOAN_ASSET,
+						input_amount: INIT_COLLATERAL + PRINCIPAL,
+						remaining_input_amount: INIT_COLLATERAL + PRINCIPAL,
+						accumulated_output_amount: 0,
+						swap_type: SwapRequestType::Regular {
+							output_action: SwapOutputAction::CreditLendingPool {
+								swap_type: LendingSwapType::Liquidation {
+									borrower_id: BORROWER,
+									loan_id: LOAN_ID
+								},
+							},
+						},
+						broker_fees: Default::default(),
+						origin: SwapOrigin::Internal,
+						price_limits_and_expiry: Some(SOFT_SWAP_PRICE_LIMIT),
+						dca_params: Some(DcaParameters { number_of_chunks: 3, chunk_interval: 1 }),
+					}
+				)])
+			);
+
+			// Finish the swap
+			MockSwapRequestHandler::<Test>::set_swap_request_progress(
```

Comment:
Nit: we use `process_loan_swap_outcome` for completing swaps and `set_swap_request_progress` for partial execution. Your way happens to work but it doesn't represent how these functions will be called in practice.